docs/principles-for-package-repository-security.md

```diff
@@ -0,0 +1,105 @@
+
+The roadmap of security capabilities can then be used by package repositories to assess gaps, put together fundable improvement lists ([like Python Packaging WG](https://github.com/psf/fundable-packaging-improvements/blob/master/FUNDABLES.md)) or write specific grant proposals that reference this guidance.
+
+## Taxonomy of Package Repositories
+
+Security capabilities will differ based on the services that the package repository offers. For instance, if the package repository has user accounts, the package repository will need to enforce authentication securely. In this section, we lay out the various relevant aspects of package repositories.
+
+Does the package repository ...
+
+... have user accounts?
+- Yes: then you have to manage things like authentication and account recovery. For example: PyPI
+- No: For example: {index, proxy, sum}.golang.org
+
+... accept built packages, do builds on behalf of users, or only host source code?
+- Accept built packages: then you have to provide a way to link packages back to source code and build instructions. For example: npm
+- Builds on behalf of users: For example, Homebrew
+- Only host source code: For example: {index, proxy, sum}.golang.org
+
+## Security Capabilities of Package Repositories
+
+We define 4 levels of security maturity:
+- Level 0 is defined as having very little security maturity.
+- Level 1 is defined as having basic security maturity. Package repositories support basic security features like multi-factor authentication, allow security researchers to report vulnerabilities. All package management ecosystems should be working towards this level.
+- Level 2 is defined as having moderate security.
+- Level 3 is defined as having advanced security. These items are more aspirational, especially for smaller package management ecosystems.
+
+### Authentication
+
+Applies to: package repositories that have user accounts.
+
+- Level 1
+  - The package repository requires users to verify their email address. This ensures that if a user loses access to their account, they are able to recover it.
+  - The package repository documents their account recovery policy.
+  - The package repository supports multi-factor authentication (MFA) via (at minimum) TOTP.
+  - The package repository notifies maintainers via email for critical account security changes, such as password changes or disabling multi-factor authentication. This helps users detect if their account has been compromised.
+  - The package repository implements account security measures like brute force prevention (even with 2FA attempts)
+- Level 2
+  - To prevent domain resurrection for account takeover via the recovery process, the package repository detects abandoned email domains. This may look like doing a WHOIS lookup on all registered email domains, and locking accounts for which the domain is abandoned.
```

Comment:
This first requires email addresses and then locks people out of their account for losing them.
Some people just want to publish software pseudonymously and use disposable email addresses on sites that force email verification. Normally username+password are sufficient for their purposes and the email is more of a burden.

Instead of making the loss of an email address even worse by also locking people out of an account it's better to only invalidate the email address and let the username+password(+optional additional factors) keep working for as many years as they can maintain their password managers.

And as with my previous comment, instead of burdening everyone it would be more gentle to indicate the level of individual projects without ratcheting up the barrier of entry for everyone using the repository.

Reply:
On both GitHub and npm, being able to recover access to your account with an email address is a **very** popular feature. I understand your preference, but balancing conflicting requests is one of the challenges of operating systems at scale.

Reply:
Do note that this document, as with everything this working group does, is a recommendation, not a requirement. From our README:

> The working group is not a governing body and does not create binding obligations on members

If a package repository decides that a different path is better, that is entirely their choice!

Reply:
Public recommendations but reputable organizations have the tendency of being turned into non-negotiable requirements by checkbox-tickers. Therefore careful wording is crucial.

I'm specifically having an issue with the "locking account" wording in this section. Instead if should say something like "invalidate that particular mail address". Some repos have multiple recovery options, support multiple emails or something like that. And having an non-recoverable account into one can at least still log in is still preferable over getting it locked.

Reply:
That's fair! I have updated the wording to reflect that what we really want is to remove the ability to recover the account from an abandoned email domain.